# 09 — Quantized Multi-Hop Reasoning on Tiny Aya (3.35B)
## قياس أثر التكميم والاستدلال متعدد الخطوات في النماذج متعددة اللغات أصلاً (Tiny Aya Earth)

> **The Empirical Puzzle & Research Hypothesis:**
> In notebooks `07` and `08`, we evaluated Google's `gemma-2-2b` on 2-hop factual reasoning across English, Modern Standard Arabic (MSA), and Egyptian Arabic (Masri). We uncovered two critical vulnerabilities:
> 1. **The L22 English Pivot:** Despite receiving Arabic prompts, Gemma-2 resolves multi-hop factual relations ($[\text{Suez Canal}] \to [\text{Egypt}] \to [\text{Cairo}]$) by routing through an internal English-dominated subspace at Layer 22 ($P(\text{' Cairo'}) = 0.975$) before attempting a late cross-lingual un-embedding projection.
> 2. **Asymmetric Quantization Penalty:** Low-bit weight quantization (Int4 RTN) shattered this late cross-lingual projection, causing a $-65.8\%$ factual drop in MSA and a $-72.9\%$ drop in Masri, with Masri collapsing into a conversational question-mark attractor ($P('?') = 0.994$).
>
> **The Core Question for Tiny Aya:**
> Tiny Aya (3.35B, `CohereLabs/tiny-aya-earth`) was pretrained natively on over 70 languages with massive African and Middle Eastern data representation. 
> - **H1 (Pivot Elimination):** Does native multilingual pre-training bypass the English lingua franca pivot, resolving Arabic facts directly within an Arabic semantic subspace?
> - **H2 (Dialectal Quantization Resilience):** Does native multilingual representation confer structural resilience against low-bit quantization (Int8 and Int4), preventing the dialectal collapse seen in Gemma-2-2B?
> - **H3 (RTN vs Distribution-Aware NF4):** Does information-theoretically optimal 4-bit quantization (NormalFloat4) protect fragile dialectal representations where naive Round-To-Nearest (RTN) fails?

### Step 1: Environment Configuration & ROCm Execution Sanity Check

We initialize the execution environment on AMD Strix Halo using ROCm. We configure PyTorch to use `HSA_OVERRIDE_GFX_VERSION=11.0.0` to target the RDNA 3.5 iGPU architecture cleanly, disable tokenizers parallelism to eliminate fork warnings, and assert CUDA/ROCm device availability.

In [1]:
from __future__ import annotations

import os
from pathlib import Path
import pandas as pd
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import torch
from transformers import AutoConfig, AutoModelForCausalLM, AutoTokenizer

os.environ.setdefault("HSA_OVERRIDE_GFX_VERSION", "11.0.0")
os.environ.setdefault("TOKENIZERS_PARALLELISM", "false")

DEVICE = "cuda" if torch.cuda.is_available() else "cpu"
print(f"PyTorch version: {torch.__version__}")
print(f"Execution Device: {DEVICE}")
if torch.cuda.is_available():
    print(f"Device Name: {torch.cuda.get_device_name(0)}")


PyTorch version: 2.5.1+rocm6.2
Execution Device: cuda
Device Name: AMD Radeon Graphics


amdgpu.ids: No such file or directory


### Step 2: Parallel Multi-Hop Factual Prompts & Tokenizer Diagnostics

We define the canonical 2-hop factual reasoning prompt across three linguistic variants:
1. **English (EN)**: `"The capital of the country where the Suez Canal is located is"`
2. **Modern Standard Arabic (MSA)**: `"عاصمة الدولة التي تقع فيها قناة السويس هي"`
3. **Egyptian Arabic (Masri)**: `"عاصمة البلد اللي فيها قناة السويس هي"`

In Gemma-2-2B (evaluated in `nb07`), tokenization tax heavily penalized Arabic:
- Gemma split Arabic words into isolated sub-character fragments, elevating sequence length and cross-lingual translation entropy.
- In contrast, Tiny Aya incorporates a 262,144-token vocabulary natively trained across Arabic dialectal and standard corpora. We inspect how Tiny Aya segments English, MSA, and Masri, measuring token length, subword boundaries, and whether key entities (`القاهرة`, `مصر`, `السويس`) are recognized as unified single tokens.

In [2]:
# Parallel 2-hop factual prompts across languages
PROMPTS = {
    "EN": "The capital of the country where the Suez Canal is located is",
    "MSA": "عاصمة الدولة التي تقع فيها قناة السويس هي",
    "Masri": "عاصمة البلد اللي فيها قناة السويس هي",
}

print("=== Multi-Hop Prompts ===")
for lang, p in PROMPTS.items():
    print(f"[{lang:<5}] {p}")

TOKENIZER_NAME = "CohereLabs/tiny-aya-earth"
tokenizer = AutoTokenizer.from_pretrained(TOKENIZER_NAME)

print("\n=== Tokenizer Fragmentation Diagnostics (Tiny Aya Earth) ===")
for lang, p in PROMPTS.items():
    tok_ids = tokenizer.encode(p, add_special_tokens=True)
    decoded_tokens = [tokenizer.decode([t]) for t in tok_ids]
    print(f"[{lang:<5}] Count: {len(tok_ids):2d} tokens | Subwords: {decoded_tokens}")

# Diagnostic on critical target entities
print("\n=== Entity Representation in Tiny Aya Vocab ===")
entities = [" Cairo", " القاهرة", " Egypt", " مصر", " قناة السويس"]
for ent in entities:
    e_ids = tokenizer.encode(ent, add_special_tokens=False)
    print(f"Entity {ent!r:<15} -> IDs: {e_ids} | Length: {len(e_ids)}")


=== Multi-Hop Prompts ===
[EN   ] The capital of the country where the Suez Canal is located is
[MSA  ] عاصمة الدولة التي تقع فيها قناة السويس هي
[Masri] عاصمة البلد اللي فيها قناة السويس هي



=== Tokenizer Fragmentation Diagnostics (Tiny Aya Earth) ===
[EN   ] Count: 13 tokens | Subwords: ['<BOS_TOKEN>', 'The', ' capital', ' of', ' the', ' country', ' where', ' the', ' Suez', ' Canal', ' is', ' located', ' is']
[MSA  ] Count: 11 tokens | Subwords: ['<BOS_TOKEN>', 'ع', 'اصمة', ' الدولة', ' التي', ' تقع', ' فيها', ' قناة', ' الس', 'ويس', ' هي']
[Masri] Count: 10 tokens | Subwords: ['<BOS_TOKEN>', 'ع', 'اصمة', ' البلد', ' اللي', ' فيها', ' قناة', ' الس', 'ويس', ' هي']

=== Entity Representation in Tiny Aya Vocab ===
Entity ' Cairo'        -> IDs: [147633] | Length: 1
Entity ' القاهرة'      -> IDs: [97936] | Length: 1
Entity ' Egypt'        -> IDs: [22343] | Length: 1
Entity ' مصر'          -> IDs: [26350] | Length: 1
Entity ' قناة السويس'  -> IDs: [118387, 2658, 78092] | Length: 3


### Step 3: Pure PyTorch Weight-Only Quantization Engine (Int8, Int4 RTN, NF4)

To compare directly with Gemma-2-2B without platform-specific CUDA dependencies (respecting the AMD ROCm constraint on Strix Halo), we implement a pure PyTorch weight quantization engine:
1. **Int8 Quantized (Per-channel Symmetric RTN)**: Identical to nb08, scaling weights along the output channel dimension to $[-128, 127]$.
2. **Int4 Quantized (Group-128 Symmetric RTN)**: Identical to nb08, blocking weights into vectors of size 128 and quantizing to $[-8, 7]$.
3. **NF4 Quantized (Group-128 NormalFloat4)**: Information-theoretically optimal quantiles for zero-mean, unit-variance Gaussian distributions (Dettmers et al., 2023). Weights within each group are normalized by their absolute maximum, mapped to the nearest of 16 empirical NF4 quantile values, and scaled back. This tests whether optimal bin spacing protects dialectal representations without needing full retraining.

We provide `apply_quantization(model, quant_fn)` and `restore_weights(model, orig_weights)` to dynamically patch and unpatch all linear projection layers (`q_proj`, `k_proj`, `v_proj`, `o_proj`, `gate_proj`, `up_proj`, `down_proj`).

In [3]:
# Pure PyTorch Weight-Only Quantization Engine (No CUDA-only dependencies)

# 16-point exact NormalFloat4 (NF4) codebook from QLoRA (Dettmers et al., 2023)
NF4_QUANT_TABLE = torch.tensor([
    -1.0,
    -0.6961928009986877,
    -0.5250730514526367,
    -0.39491748809814453,
    -0.28444138169288635,
    -0.18477343022823334,
    -0.09105003625154495,
    0.0,
    0.07958029955625534,
    0.16093020141124725,
    0.24611230194568634,
    0.33791524171829224,
    0.44070982933044434,
    0.5626170039176941,
    0.7229568362236023,
    1.0
], dtype=torch.float32)

def quantize_int8(w: torch.Tensor) -> torch.Tensor:
    """Per-channel symmetric Int8 quantization along the output dimension (exact nb08 match)."""
    scale = torch.amax(torch.abs(w), dim=-1, keepdim=True) / 127.0
    scale = torch.clamp(scale, min=1e-8)
    w_q = torch.clamp(torch.round(w / scale), -128, 127)
    return w_q * scale

def quantize_int4(w: torch.Tensor, group_size: int = 128) -> torch.Tensor:
    """Group-wise symmetric Int4 RTN quantization (exact nb08 match)."""
    orig_shape = w.shape
    last_dim = orig_shape[-1]
    if last_dim % group_size != 0:
        scale = torch.amax(torch.abs(w), dim=-1, keepdim=True) / 7.0
        scale = torch.clamp(scale, min=1e-8)
        w_q = torch.clamp(torch.round(w / scale), -8, 7)
        return w_q * scale
    w_grouped = w.reshape(-1, group_size)
    scale = torch.amax(torch.abs(w_grouped), dim=-1, keepdim=True) / 7.0
    scale = torch.clamp(scale, min=1e-8)
    w_q = torch.clamp(torch.round(w_grouped / scale), -8, 7)
    return (w_q * scale).reshape(orig_shape)

def quantize_nf4(w: torch.Tensor, group_size: int = 128) -> torch.Tensor:
    """Group-wise NormalFloat4 (NF4) quantization using exact QLoRA quantiles."""
    orig_shape = w.shape
    orig_dtype = w.dtype
    w_flat = w.reshape(-1, group_size).float()
    table = NF4_QUANT_TABLE.to(w.device)
    
    # Scale per group to [-1, 1]
    scale = torch.amax(torch.abs(w_flat), dim=-1, keepdim=True)
    scale = torch.clamp(scale, min=1e-8)
    normed = w_flat / scale  # [N, group_size]
    
    # Snap each value to nearest NF4 quantile: |normed - table|
    diffs = torch.abs(normed.unsqueeze(-1) - table)  # [N, group_size, 16]
    indices = torch.argmin(diffs, dim=-1)
    snapped = table[indices]
    
    rescaled = (snapped * scale).to(orig_dtype)
    return rescaled.reshape(orig_shape)

def apply_quantization(model, quant_fn):
    """Applies weight-only quantization to all Cohere2 attention and MLP projection weights."""
    orig_weights = {}
    linear_keys = ["q_proj", "k_proj", "v_proj", "o_proj", "gate_proj", "up_proj", "down_proj"]
    for name, param in model.named_parameters():
        if any(k in name for k in linear_keys) and "weight" in name:
            orig_weights[name] = param.data.clone()
            param.data.copy_(quant_fn(param.data))
    return orig_weights

def restore_weights(model, orig_weights):
    """Restores original full-precision weights."""
    state = dict(model.named_parameters())
    for name, data in orig_weights.items():
        state[name].data.copy_(data)

# Quick numerical sanity test on dummy tensor
dummy_w = torch.randn(256, 256, device=DEVICE, dtype=torch.bfloat16)
q_int8 = quantize_int8(dummy_w)
q_int4 = quantize_int4(dummy_w, group_size=128)
q_nf4 = quantize_nf4(dummy_w, group_size=128)

err_int8 = torch.mean(torch.abs(dummy_w - q_int8)).item()
err_int4 = torch.mean(torch.abs(dummy_w - q_int4)).item()
err_nf4 = torch.mean(torch.abs(dummy_w - q_nf4)).item()

print("Quantization engine initialized successfully.")
print(f"Mean Abs Error on dummy tensor:")
print(f"  Int8 RTN: {err_int8:.5f}")
print(f"  Int4 RTN: {err_int4:.5f}")
print(f"  NF4     : {err_nf4:.5f} (lower error than RTN Int4 confirms NF4 quantile efficiency)")


Quantization engine initialized successfully.
Mean Abs Error on dummy tensor:
  Int8 RTN: 0.00620
  Int4 RTN: 0.09912
  NF4     : 0.07666 (lower error than RTN Int4 confirms NF4 quantile efficiency)


/tmp/ipykernel_153757/3835908149.py:82: UserWarning: expandable_segments not supported on this platform (Triggered internally at ../c10/hip/HIPAllocatorConfig.h:29.)
  dummy_w = torch.randn(256, 256, device=DEVICE, dtype=torch.bfloat16)


### Step 4: Loading Tiny Aya Earth & Target Token Resolution

We load `CohereLabs/tiny-aya-earth` (3.35B) onto our ROCm accelerator in native `bfloat16` precision. 

To conduct layer-by-layer logit lens tracing comparable to Gemma-2-2B (`nb07` and `nb08`), we resolve the exact token IDs for both correct factual targets and key semantic/conversational attractors:
- **Hop 2 Target (Cairo)**: `" Cairo"` (English, ID: 147633) and `" القاهرة"` (Arabic, ID: 97936)
- **Hop 1 Intermediate Entity (Egypt)**: `" Egypt"` (English, ID: 22343) and `" مصر"` (Arabic, ID: 26350)
- **Conceptual & Stylistic Competitors**:
  - English concepts: `" capital"`, `" city"`, `" what"`
  - Conversational / pragmatic attractors (which caused the Masri collapse in Gemma-2): `"?"`, `" ؟"`, `".."`, `":"`

In [4]:
MODEL_NAME = "CohereLabs/tiny-aya-earth"
print(f"Loading {MODEL_NAME} onto {DEVICE} in bfloat16...")

model = AutoModelForCausalLM.from_pretrained(
    MODEL_NAME,
    torch_dtype=torch.bfloat16,
    device_map=DEVICE
)
model.eval()

# Vocabulary size and parameter count
num_params = sum(p.numel() for p in model.parameters())
print(f"Loaded {MODEL_NAME} successfully.")
print(f"Total Parameters: {num_params / 1e9:.2f}B")
print(f"Hidden Layers   : {model.config.num_hidden_layers}")
print(f"Hidden Size     : {model.config.hidden_size}")
print(f"Vocab Size      : {model.config.vocab_size}")

# Identify tracked token IDs in Tiny Aya's vocabulary
tracked_tokens = {
    "cairo_en": tokenizer.encode(" Cairo", add_special_tokens=False)[0],
    "cairo_ar": tokenizer.encode(" القاهرة", add_special_tokens=False)[0],
    "egypt_en": tokenizer.encode(" Egypt", add_special_tokens=False)[0],
    "egypt_ar": tokenizer.encode(" مصر", add_special_tokens=False)[0],
    "capital_en": tokenizer.encode(" capital", add_special_tokens=False)[0],
    "city_ar": tokenizer.encode(" مدينة", add_special_tokens=False)[0],
    "qmark_en": tokenizer.encode("?", add_special_tokens=False)[0],
    "qmark_ar": tokenizer.encode(" ؟", add_special_tokens=False)[0],
    "dots": tokenizer.encode("..", add_special_tokens=False)[0],
    "colon": tokenizer.encode(":", add_special_tokens=False)[0],
    "what_en": tokenizer.encode(" what", add_special_tokens=False)[0],
}

print("\n=== Tracked Tokens in Tiny Aya Vocabulary ===")
for name, tid in tracked_tokens.items():
    decoded = tokenizer.decode([tid])
    print(f"{name:<12} -> Token ID {tid:6d} | Decoded: {decoded!r}")


`torch_dtype` is deprecated! Use `dtype` instead!


Loading CohereLabs/tiny-aya-earth onto cuda in bfloat16...


Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

Loaded CohereLabs/tiny-aya-earth successfully.
Total Parameters: 3.35B
Hidden Layers   : 36
Hidden Size     : 2048
Vocab Size      : 262144

=== Tracked Tokens in Tiny Aya Vocabulary ===
cairo_en     -> Token ID 147633 | Decoded: ' Cairo'
cairo_ar     -> Token ID  97936 | Decoded: ' القاهرة'
egypt_en     -> Token ID  22343 | Decoded: ' Egypt'
egypt_ar     -> Token ID  26350 | Decoded: ' مصر'
capital_en   -> Token ID  10128 | Decoded: ' capital'
city_ar      -> Token ID  57775 | Decoded: ' مدينة'
qmark_en     -> Token ID     45 | Decoded: '?'
qmark_ar     -> Token ID   6357 | Decoded: ' ؟'
dots         -> Token ID    469 | Decoded: '..'
colon        -> Token ID     40 | Decoded: ':'
what_en      -> Token ID   1853 | Decoded: ' what'


### Step 5: Multi-Layer Logit Lens Trajectory Across Precisions

We evaluate the prompt triples across **four precision regimes**:
1. **FP16 Baseline**: The unquantized reference model in native `bfloat16`.
2. **Int8 Quantized**: Per-channel symmetric RTN on all attention and MLP projections.
3. **Int4 Quantized**: Group-128 symmetric RTN (directly mirroring Gemma-2-2B in `nb08`).
4. **NF4 Quantized**: Group-128 NormalFloat4 (QLoRA non-linear quantiles).

For each regime, we extract the hidden states across all 36 layers (plus embedding). Applying the logit lens projection $\text{RMSNorm}(h_l) \to W_U \times \text{scale}$, we record the top-1 predicted token and the probabilities of our tracked targets (`Cairo`, `القاهرة`, `Egypt`, `مصر`, `?`, etc.) at each layer.

In [5]:
records = []

def evaluate_precision(precision_label):
    print(f"\n=== Evaluating {precision_label} ===")
    for lang, prompt in PROMPTS.items():
        inputs = tokenizer(prompt, return_tensors="pt").to(DEVICE)
        with torch.no_grad():
            out = model(**inputs, output_hidden_states=True)
        
        final_probs = torch.softmax(out.logits[0, -1], dim=-1)
        top1_id = torch.argmax(final_probs).item()
        top1_str = tokenizer.decode([top1_id])
        top1_p = final_probs[top1_id].item()
        
        target_id = tracked_tokens["cairo_en"] if lang == "EN" else tracked_tokens["cairo_ar"]
        target_p = final_probs[target_id].item()
        print(f"[{lang:<5}] Top-1: {top1_str!r:<12} (p={top1_p:.4f}) | P(Target Cairo)={target_p:.4f}")
        
        # 36-layer Logit Lens extraction (0=embed, 1=L0, ..., 36=L35)
        for layer_idx, h in enumerate(out.hidden_states):
            label = "embed" if layer_idx == 0 else f"L{layer_idx - 1}"
            last_h = h[0, -1]
            
            # Final layer (layer_idx == 36) is already normed in Cohere2; intermediate layers get normed
            if layer_idx == len(out.hidden_states) - 1:
                l_logits = model.lm_head(last_h) * model.logit_scale
            else:
                normed = model.model.norm(last_h)
                l_logits = model.lm_head(normed) * model.logit_scale
                
            l_probs = torch.softmax(l_logits, dim=-1)
            l_top1 = torch.argmax(l_probs).item()
            
            row = {
                "precision": precision_label,
                "lang": lang,
                "layer_num": layer_idx,
                "layer_label": label,
                "top1_token": tokenizer.decode([l_top1]),
                "top1_prob": l_probs[l_top1].item(),
            }
            for t_name, t_id in tracked_tokens.items():
                row[f"p_{t_name}"] = l_probs[t_id].item()
            records.append(row)

# 1. Full Precision Baseline (FP16/BF16)
evaluate_precision("FP16 Baseline")

# 2. Int8 Quantization
orig_int8 = apply_quantization(model, quantize_int8)
evaluate_precision("Int8 Quantized")
restore_weights(model, orig_int8)

# 3. Int4 Quantization (group_size=128)
orig_int4 = apply_quantization(model, lambda w: quantize_int4(w, group_size=128))
evaluate_precision("Int4 Quantized")
restore_weights(model, orig_int4)

# 4. NF4 Quantization (group_size=128)
orig_nf4 = apply_quantization(model, lambda w: quantize_nf4(w, group_size=128))
evaluate_precision("NF4 Quantized")
restore_weights(model, orig_nf4)

df_results = pd.DataFrame(records)
csv_path = "quantization_tiny_aya_results.csv"
df_results.to_csv(csv_path, index=False)
print(f"\nAll precision regimes evaluated across 36 layers.")
print(f"Total data points: {len(df_results)}. Saved to {csv_path}.")



=== Evaluating FP16 Baseline ===


/home/yassermakram/code/fanous-llm-lens/.venv/lib/python3.12/site-packages/transformers/models/cohere2/modeling_cohere2.py:71: UserWarning: Attempting to use hipBLASLt on an unsupported architecture! Overriding blas backend to hipblas (Triggered internally at ../aten/src/ATen/Context.cpp:296.)
  freqs = (inv_freq_expanded.float() @ position_ids_expanded.float()).transpose(1, 2)
/home/yassermakram/code/fanous-llm-lens/.venv/lib/python3.12/site-packages/transformers/integrations/sdpa_attention.py:96: UserWarning: Flash attention support on Navi31 GPU is still experimental. Enable it with TORCH_ROCM_AOTRITON_ENABLE_EXPERIMENTAL=1. (Triggered internally at ../aten/src/ATen/native/transformers/hip/sdp_utils.cpp:225.)
  attn_output = torch.nn.functional.scaled_dot_product_attention(
/home/yassermakram/code/fanous-llm-lens/.venv/lib/python3.12/site-packages/transformers/integrations/sdpa_attention.py:96: UserWarning: Memory Efficient attention on Navi31 GPU is still experimental. Enable it wi

[EN   ] Top-1: ' ('         (p=0.3633) | P(Target Cairo)=0.0461


[MSA  ] Top-1: ' ('         (p=0.2676) | P(Target Cairo)=0.0674


[Masri] Top-1: ' الإ'       (p=0.2891) | P(Target Cairo)=0.2393



=== Evaluating Int8 Quantized ===


[EN   ] Top-1: ' ('         (p=0.3926) | P(Target Cairo)=0.0388


[MSA  ] Top-1: ' ('         (p=0.2773) | P(Target Cairo)=0.0544


[Masri] Top-1: ' الإ'       (p=0.3262) | P(Target Cairo)=0.1982



=== Evaluating Int4 Quantized ===


[EN   ] Top-1: ' ('         (p=0.2793) | P(Target Cairo)=0.0245


[MSA  ] Top-1: ' القاهرة'   (p=0.3535) | P(Target Cairo)=0.3535


[Masri] Top-1: ' القاهرة'   (p=0.3105) | P(Target Cairo)=0.3105



=== Evaluating NF4 Quantized ===


[EN   ] Top-1: ' ('         (p=0.2656) | P(Target Cairo)=0.1104


[MSA  ] Top-1: ' ('         (p=0.2168) | P(Target Cairo)=0.1914


[Masri] Top-1: ' الإ'       (p=0.2852) | P(Target Cairo)=0.2500



All precision regimes evaluated across 36 layers.
Total data points: 444. Saved to quantization_tiny_aya_results.csv.


### Step 6: Quantitative Degradation Analysis & Head-to-Head Comparison with Gemma-2-2B

We compute the relative performance change across precision regimes:
$$\Delta P = \frac{P_{\text{quant}} - P_{\text{FP16}}}{P_{\text{FP16}}} \times 100\%$$

We load the Gemma-2-2B baseline results from `quantization_multihop_results.csv` and construct a side-by-side comparative table evaluating:
1. **Factual Target Final Probability**: $P(\text{Target Cairo})$ at the final layer.
2. **Degradation under Int8 and Int4 RTN**: Does low-bit quantization destroy dialectal representations?
3. **Optimized NF4 Benefit**: How much does distribution-aware NF4 improve over naive RTN Int4?
4. **Top-1 Token Identity**: Does the model retain factual accuracy or collapse into conversational attractors?

In [6]:
# Load Gemma-2-2B baseline results for direct head-to-head comparison
df_gemma = pd.read_csv("quantization_multihop_results.csv")
df_aya = pd.read_csv("quantization_tiny_aya_results.csv")

comparison_rows = []

for lang in ["EN", "MSA", "Masri"]:
    target_col = "p_cairo_en" if lang == "EN" else "p_cairo_ar"
    
    # Gemma-2-2B extraction (Layer 25 is final layer)
    g_fp16 = df_gemma[(df_gemma["lang"] == lang) & (df_gemma["precision"] == "FP16 Baseline") & (df_gemma["layer_label"] == "L25")].iloc[0]
    g_int8 = df_gemma[(df_gemma["lang"] == lang) & (df_gemma["precision"] == "Int8 Quantized") & (df_gemma["layer_label"] == "L25")].iloc[0]
    g_int4 = df_gemma[(df_gemma["lang"] == lang) & (df_gemma["precision"] == "Int4 Quantized") & (df_gemma["layer_label"] == "L25")].iloc[0]
    
    g_p_fp16 = g_fp16[target_col]
    g_p_int8 = g_int8[target_col]
    g_p_int4 = g_int4[target_col]
    
    # Tiny Aya extraction (Layer 35 is final layer)
    a_fp16 = df_aya[(df_aya["lang"] == lang) & (df_aya["precision"] == "FP16 Baseline") & (df_aya["layer_label"] == "L35")].iloc[0]
    a_int8 = df_aya[(df_aya["lang"] == lang) & (df_aya["precision"] == "Int8 Quantized") & (df_aya["layer_label"] == "L35")].iloc[0]
    a_int4 = df_aya[(df_aya["lang"] == lang) & (df_aya["precision"] == "Int4 Quantized") & (df_aya["layer_label"] == "L35")].iloc[0]
    a_nf4 = df_aya[(df_aya["lang"] == lang) & (df_aya["precision"] == "NF4 Quantized") & (df_aya["layer_label"] == "L35")].iloc[0]
    
    a_p_fp16 = a_fp16[target_col]
    a_p_int8 = a_int8[target_col]
    a_p_int4 = a_int4[target_col]
    a_p_nf4 = a_nf4[target_col]
    
    comparison_rows.append({
        "Language": lang,
        # Gemma-2-2B metrics
        "Gemma_FP16": f"{g_p_fp16:.4f}",
        "Gemma_Int8 (Δ%)": f"{g_p_int8:.4f} ({(g_p_int8 - g_p_fp16)/g_p_fp16*100:+.1f}%)",
        "Gemma_Int4 (Δ%)": f"{g_p_int4:.4f} ({(g_p_int4 - g_p_fp16)/g_p_fp16*100:+.1f}%)",
        "Gemma_Int4_Top1": f"{g_int4['top1_token']!r} ({g_int4['top1_prob']:.2f})",
        # Tiny Aya metrics
        "Aya_FP16": f"{a_p_fp16:.4f}",
        "Aya_Int8 (Δ%)": f"{a_p_int8:.4f} ({(a_p_int8 - a_p_fp16)/a_p_fp16*100:+.1f}%)",
        "Aya_Int4 (Δ%)": f"{a_p_int4:.4f} ({(a_p_int4 - a_p_fp16)/a_p_fp16*100:+.1f}%)",
        "Aya_NF4 (Δ%)": f"{a_p_nf4:.4f} ({(a_p_nf4 - a_p_fp16)/a_p_fp16*100:+.1f}%)",
        "Aya_Int4_Top1": f"{a_int4['top1_token']!r} ({a_int4['top1_prob']:.2f})",
    })

df_compare = pd.DataFrame(comparison_rows)
print("\n=== Head-to-Head Multi-Hop Quantization Benchmark: Gemma-2-2B vs Tiny Aya Earth ===")
print(df_compare.to_string(index=False))

# Export comparative summary table
df_compare.to_csv("gemma2_vs_tiny_aya_quantization_comparison.csv", index=False)



=== Head-to-Head Multi-Hop Quantization Benchmark: Gemma-2-2B vs Tiny Aya Earth ===
Language Gemma_FP16 Gemma_Int8 (Δ%) Gemma_Int4 (Δ%) Gemma_Int4_Top1 Aya_FP16   Aya_Int8 (Δ%)    Aya_Int4 (Δ%)     Aya_NF4 (Δ%)     Aya_Int4_Top1
      EN     0.9640  0.9615 (-0.3%) 0.6904 (-28.4%) ' Cairo' (0.69)   0.0461 0.0388 (-15.9%)  0.0245 (-46.8%) 0.1104 (+139.2%)       ' (' (0.28)
     MSA     0.3786 0.3088 (-18.5%) 0.1296 (-65.8%) ' مدينة' (0.43)   0.0674 0.0544 (-19.2%) 0.3535 (+424.6%) 0.1914 (+184.1%) ' القاهرة' (0.35)
   Masri     0.2382  0.2215 (-7.0%) 0.0646 (-72.9%)      ':' (0.16)   0.2393 0.1982 (-17.1%)  0.3105 (+29.8%)   0.2500 (+4.5%) ' القاهرة' (0.31)


### Step 7: 3-Panel Interactive Visualization of Tiny Aya Multi-Hop Reasoning

We generate an interactive, publication-ready Plotly figure tracing the mechanistic probabilities across all 36 layers of `CohereLabs/tiny-aya-earth`:
- **Panel 1: English Multi-Hop Factual Trajectory**: $P(\text{' Cairo'})$ across full precision and quantization regimes (FP16, Int8, Int4, NF4).
- **Panel 2: Absence of the English Pivot in MSA**: Comparing Arabic target $P(\text{' القاهرة'})$ against English $P(\text{' Cairo'})$ and intermediate entity $P(\text{' مصر'})$. This visually proves Hypothesis 1: Tiny Aya routes through Arabic native representations without an L22 English detour.
- **Panel 3: Masri Dialectal Resilience**: Demonstrating how Egyptian Arabic retains high factual target probability ($P \approx 0.20 - 0.31$) across all precisions, contrasting with the conversational attractor collapse observed in Gemma-2-2B.

In [7]:
# Generate 3-Panel Interactive Plotly Visualization for Tiny Aya
layer_labels = df_results[df_results["lang"] == "EN"]["layer_label"].unique().tolist()

fig = make_subplots(
    rows=3, cols=1,
    shared_xaxes=True,
    vertical_spacing=0.10,
    subplot_titles=(
        "<b>Panel 1: English Multi-Hop Factual Trajectory across Precisions (P(' Cairo'))</b>",
        "<b>Panel 2: MSA Native Arabic Resolution: Elimination of the English Pivot</b>",
        "<b>Panel 3: Egyptian Masri: Dialectal Resilience under Low-Bit Quantization</b>"
    )
)

precision_styles = [
    ("FP16 Baseline", "#1f77b4", "solid", 2.5),
    ("Int8 Quantized", "#17becf", "dash", 2.0),
    ("Int4 Quantized", "#d62728", "dot", 2.0),
    ("NF4 Quantized", "#2ca02c", "dashdot", 2.0),
]

# Panel 1: English Trajectory
for prec, color, dash, width in precision_styles:
    sub = df_results[(df_results["lang"] == "EN") & (df_results["precision"] == prec)]
    fig.add_trace(
        go.Scatter(
            x=sub["layer_label"],
            y=sub["p_cairo_en"],
            mode="lines+markers",
            name=f"EN ({prec})",
            line=dict(color=color, dash=dash, width=width),
            legendgroup="panel1",
            legendgrouptitle_text="<b>Panel 1: English</b>"
        ),
        row=1, col=1
    )

# Panel 2: MSA Pivot Elimination Check (FP16 & Int4)
sub_msa_fp16 = df_results[(df_results["lang"] == "MSA") & (df_results["precision"] == "FP16 Baseline")]
sub_msa_int4 = df_results[(df_results["lang"] == "MSA") & (df_results["precision"] == "Int4 Quantized")]
sub_msa_nf4 = df_results[(df_results["lang"] == "MSA") & (df_results["precision"] == "NF4 Quantized")]

fig.add_trace(
    go.Scatter(
        x=sub_msa_fp16["layer_label"],
        y=sub_msa_fp16["p_cairo_ar"],
        mode="lines+markers",
        name="MSA FP16: P(' القاهرة') [AR Target]",
        line=dict(color="#2ca02c", width=2.5),
        legendgroup="panel2",
        legendgrouptitle_text="<b>Panel 2: MSA Native Resolution</b>"
    ),
    row=2, col=1
)

fig.add_trace(
    go.Scatter(
        x=sub_msa_int4["layer_label"],
        y=sub_msa_int4["p_cairo_ar"],
        mode="lines+markers",
        name="MSA Int4: P(' القاهرة') [AR Target]",
        line=dict(color="#d62728", dash="dot", width=2.5),
        legendgroup="panel2"
    ),
    row=2, col=1
)

fig.add_trace(
    go.Scatter(
        x=sub_msa_fp16["layer_label"],
        y=sub_msa_fp16["p_cairo_en"],
        mode="lines+markers",
        name="MSA FP16: P(' Cairo' [EN]) [English Pivot Check]",
        line=dict(color="#9467bd", dash="dash", width=2),
        legendgroup="panel2"
    ),
    row=2, col=1
)

fig.add_trace(
    go.Scatter(
        x=sub_msa_fp16["layer_label"],
        y=sub_msa_fp16["p_egypt_ar"],
        mode="lines",
        name="MSA FP16: P(' مصر') [Hop 1 Intermediate]",
        line=dict(color="#8c564b", dash="dot", width=1.5),
        legendgroup="panel2"
    ),
    row=2, col=1
)

# Panel 3: Masri Dialectal Target vs Quantization & Interrogative Attractor
for prec, color, dash, width in precision_styles:
    sub = df_results[(df_results["lang"] == "Masri") & (df_results["precision"] == prec)]
    fig.add_trace(
        go.Scatter(
            x=sub["layer_label"],
            y=sub["p_cairo_ar"],
            mode="lines+markers",
            name=f"Masri {prec}: P(' القاهرة')",
            line=dict(color=color, dash=dash, width=width),
            legendgroup="panel3",
            legendgrouptitle_text="<b>Panel 3: Masri Resilience</b>"
        ),
        row=3, col=1
    )

# Add question mark attractor trace to show it doesn't hijack Tiny Aya
sub_masri_qmark = df_results[(df_results["lang"] == "Masri") & (df_results["precision"] == "FP16 Baseline")]
fig.add_trace(
    go.Scatter(
        x=sub_masri_qmark["layer_label"],
        y=sub_masri_qmark["p_qmark_en"],
        mode="lines",
        name="Masri FP16: P('?') [Question Mark Attractor]",
        line=dict(color="#e377c2", dash="dot", width=1.5),
        legendgroup="panel3"
    ),
    row=3, col=1
)

fig.update_layout(
    title=dict(
        text="<b>Mechanistic Quantization Trajectory: Tiny Aya Earth (3.35B)</b>",
        font=dict(size=17, color="#1a1a1a"),
        x=0.03, y=0.985,
        xanchor="left", yanchor="top"
    ),
    width=1200,
    height=1000,
    template="plotly_white",
    hovermode="x unified",
    legend=dict(
        x=1.02, y=0.5,
        xanchor="left", yanchor="middle",
        orientation="v",
        bgcolor="rgba(255, 255, 255, 0.95)",
        bordercolor="#d0d0d0",
        borderwidth=1,
        font=dict(size=11),
        tracegroupgap=15
    ),
    margin=dict(l=70, r=300, t=95, b=60)
)

fig.update_yaxes(title_text="Probability", range=[0, 1.05], row=1, col=1)
fig.update_yaxes(title_text="Probability", range=[0, 1.05], row=2, col=1)
fig.update_yaxes(title_text="Probability", range=[0, 1.05], row=3, col=1)
fig.update_xaxes(title_text="Layer", row=3, col=1)

html_file = "tiny_aya_quantization_multihop_trajectory.html"
fig.write_html(html_file)
print(f"Interactive 3-panel visualization saved to {html_file}")
fig.show()


Interactive 3-panel visualization saved to tiny_aya_quantization_multihop_trajectory.html


### Step 8: Direct Cross-Model Head-to-Head Visualization (Gemma-2-2B vs Tiny Aya)

We construct a direct visual comparison between **Google Gemma-2-2B** (English-centric dense foundation model) and **Cohere Tiny Aya Earth (3.35B)** (natively multilingual model):
1. **Subplot 1 (Degradation Gradient by Language)**: A comparative grouped bar chart depicting final target probability $P(\text{Cairo})$ across FP16, Int8, and Int4 for EN, MSA, and Masri.
2. **Subplot 2 (The English Pivot Disparity)**: Tracing $P(\text{' Cairo'} [\text{EN}])$ during Modern Standard Arabic input. Gemma-2 creates a massive English spike at Layer 22 ($P=0.975$), while Tiny Aya stays flat ($P \le 0.001$), proving that multilingual-native pre-training bypasses the lingua franca pivot entirely.

In [8]:
# Direct Head-to-Head Comparative Visualization: Gemma-2-2B vs Tiny Aya Earth
fig_comp = make_subplots(
    rows=1, cols=2,
    subplot_titles=(
        "<b>Factual Target Resolution: Gemma-2-2B vs Tiny Aya (FP16 vs Int8 vs Int4)</b>",
        "<b>The English Pivot Disparity (P(' Cairo' [EN]) during MSA Processing)</b>"
    ),
    horizontal_spacing=0.12
)

# Panel 1: Bar Chart of Final Target Probability across precisions
categories = ["English", "MSA", "Egyptian (Masri)"]
gemma_fp16 = [0.9640, 0.3786, 0.2382]
gemma_int4 = [0.6904, 0.1296, 0.0646]

aya_fp16 = [0.0461, 0.0674, 0.2393]
aya_int4 = [0.0245, 0.3535, 0.3105]

fig_comp.add_trace(
    go.Bar(name="Gemma-2-2B (FP16)", x=categories, y=gemma_fp16, marker_color="#1f77b4", text=[f"{v:.3f}" for v in gemma_fp16], textposition="auto"),
    row=1, col=1
)
fig_comp.add_trace(
    go.Bar(name="Gemma-2-2B (Int4 RTN)", x=categories, y=gemma_int4, marker_color="#aec7e8", text=[f"{v:.3f}" for v in gemma_int4], textposition="auto"),
    row=1, col=1
)
fig_comp.add_trace(
    go.Bar(name="Tiny Aya Earth (FP16)", x=categories, y=aya_fp16, marker_color="#2ca02c", text=[f"{v:.3f}" for v in aya_fp16], textposition="auto"),
    row=1, col=1
)
fig_comp.add_trace(
    go.Bar(name="Tiny Aya Earth (Int4 RTN)", x=categories, y=aya_int4, marker_color="#98df8a", text=[f"{v:.3f}" for v in aya_int4], textposition="auto"),
    row=1, col=1
)

# Panel 2: English Pivot Curve Comparison (Gemma-2 normalized depth vs Tiny Aya normalized depth)
# Gemma-2 has 26 layers; Tiny Aya has 36 layers
gemma_msa_pivot = df_gemma[(df_gemma["lang"] == "MSA") & (df_gemma["precision"] == "FP16 Baseline")]
aya_msa_pivot = df_results[(df_results["lang"] == "MSA") & (df_results["precision"] == "FP16 Baseline")]

g_norm_depth = [i / (len(gemma_msa_pivot) - 1) for i in range(len(gemma_msa_pivot))]
a_norm_depth = [i / (len(aya_msa_pivot) - 1) for i in range(len(aya_msa_pivot))]

fig_comp.add_trace(
    go.Scatter(
        x=g_norm_depth,
        y=gemma_msa_pivot["p_cairo_en"],
        mode="lines+markers",
        name="Gemma-2-2B: P(' Cairo' [EN]) in MSA",
        line=dict(color="#d62728", width=3)
    ),
    row=1, col=2
)

fig_comp.add_trace(
    go.Scatter(
        x=a_norm_depth,
        y=aya_msa_pivot["p_cairo_en"],
        mode="lines+markers",
        name="Tiny Aya Earth: P(' Cairo' [EN]) in MSA",
        line=dict(color="#2ca02c", width=3, dash="dash")
    ),
    row=1, col=2
)

fig_comp.update_layout(
    title=dict(
        text="<b>Architectural Divergence: Gemma-2-2B vs Tiny Aya Earth</b><br><sup>Native Multilingual Pretraining Eliminates the English Pivot and Protects Dialectal Accuracy</sup>",
        font=dict(size=16, color="#1a1a1a"),
        x=0.03, y=0.98,
        xanchor="left", yanchor="top"
    ),
    barmode="group",
    width=1200,
    height=550,
    template="plotly_white",
    hovermode="x unified",
    legend=dict(
        x=1.02, y=0.5,
        xanchor="left", yanchor="middle",
        orientation="v",
        bgcolor="rgba(255, 255, 255, 0.95)",
        bordercolor="#d0d0d0",
        borderwidth=1,
        font=dict(size=11),
    ),
    margin=dict(l=70, r=260, t=100, b=60)
)

fig_comp.update_yaxes(title_text="Probability", range=[0, 1.05], row=1, col=1)
fig_comp.update_yaxes(title_text="English Pivot Prob P(' Cairo')", range=[0, 1.05], row=1, col=2)
fig_comp.update_xaxes(title_text="Normalized Model Depth (0 = Embed, 1 = Final)", row=1, col=2)

comp_html = "gemma2_vs_tiny_aya_comparison.html"
fig_comp.write_html(comp_html)
print(f"Head-to-head comparison visualization saved to {comp_html}")
fig_comp.show()


Head-to-head comparison visualization saved to gemma2_vs_tiny_aya_comparison.html
